arch/arm64/core/mmu.c

```diff
@@ -806,6 +806,79 @@ static uint64_t get_tcr(int el)
 	return tcr;
 }
 
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
+
+static int flush_inv_mapped_pages(int (*cmo_fn)(void *, size_t),
+				  uint64_t *pte, int level, uint64_t base)
+{
+	int ret = 0;
+	uint64_t level_size = BIT(LEVEL_TO_VA_SIZE_SHIFT(level));
+	const uint64_t max_va_addr = (uint64_t)CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE;
+
+	if (!xlat_use_count[table_index(pte)]) {
+		return 0;
+	}
+
+	for (unsigned int i = 0; i < Ln_XLAT_NUM_ENTRIES; i++) {
+		uint64_t end, va = base + i * level_size;
+		uint8_t mem_type;
+
+		if (is_free_desc(pte[i])) {
+			continue;
+		}
+
+		if (is_table_desc(pte[i], level)) {
+			/* Move to the next translation table level */
+			ret = flush_inv_mapped_pages(cmo_fn, pte_desc_table(pte[i]), level + 1, va);
+			if (ret) {
+				break;
+			}
+			continue;
+		}
+
+		/*
+		 * From this point, this must be a leaf.
+		 * Start excluding non memory mappings
+		 */
+		mem_type = (pte[i] >> 2) & MT_TYPE_MASK;
+		if ((mem_type != MT_NORMAL) && (mem_type != MT_NORMAL_NC)) {
+			continue;
+		}
+
+		end = va + level_size - 1;
+
+		/* No intersection with RAM? */
+		if (end < CONFIG_KERNEL_VM_BASE || va >= max_va_addr) {
+			continue;
+		}
+
+		va = MAX(va, CONFIG_KERNEL_VM_BASE);
+		end = MIN(end, max_va_addr);
+
+		ret = cmo_fn(UINT_TO_POINTER(va), end - va);
+		if (ret) {
+			break;
+		}
+
+		MMU_DEBUG("Flush/inv PTE %llx at level %d: %llx-%llx\n",
+			  pte, level, va, end);
+	}
+
+	return ret;
+}
+
+int flush_inv_page_tbl_entries(int (*cmo_fn)(void *, size_t))
+{
+	uint64_t *base_xlat_table = UINT_TO_POINTER(read_ttbr0_el1());
```

Comment:
This is incorrect in the general case, as TTBR holds physical address of the table, but you are treating it as it is a virtual address. 

Also, TTBR0 holds not only base table address. Bit 0 is CnP and bits 63:48 hold ASID. So it is incorrect to cast TTBR0 value into pointer.

Why didn't you use `kernel_ptables.base_xlat_table` as rest of the code does?

Reply:
Have you seen `z_arm64_swap_ptables` function? 

And the following piece of code?

```
	domain_ptables->ttbr0 =	(((uint64_t)asid) << TTBR_ASID_SHIFT) |
		((uint64_t)(uintptr_t)domain_ptables->base_xlat_table);
```
?

And yes, looks like I was wrong: it will be wrong to use `kernel_ptables.base_xlat_table` as well.

Reply:
Pls, look at function that perform actual mapping and looking for page it is worked in the same way
https://elixir.bootlin.com/zephyr/latest/source/arch/arm64/core/mmu.c#L230

And showe another place where we modify ttbr0_el1 and write value which you mentioned here

Reply:
> And showe another place where we modify ttbr0_el1 and write value which you mentioned here

I already pointed you to `z_arm64_swap_ptables`. If you'd looked at it, you'd see that it calls `z_arm64_set_ttbr0`  which, indeed, sets TTBR0:

 ```
GTEXT(z_arm64_set_ttbr0)
SECTION_FUNC(TEXT, z_arm64_set_ttbr0)

	/* Switch the TTBR0 */
	msr	ttbr0_el1, x0
	isb

	ret
```

Could you please pay more attention at what I am writing?

Reply:
yes, I see it now
I paid attention as much as it was possible on that moment

yes you were right that may be a problem there, I'll fix it
but now there is any problem at least with the current code